subsys/net/lib/downloader/src/transports/http.c

```diff
@@ -0,0 +1,571 @@
+		LOG_HEXDUMP_DBG(dl->config.buf, len, "HTTP request");
+	}
+
+	LOG_DBG("http request:\n%s\n", dl->config.buf);
+
+	err = dl_socket_send(http->sock.fd, dl->config.buf, len);
+	if (err) {
+		LOG_ERR("Failed to send HTTP request, errno %d", errno);
+		return err;
+	}
+
+	return 0;
+}
+
+/* Returns:
+ * Number of bytes parsed on success.
+ * Negative errno on error.
+ */
+static int http_header_parse(struct downloader *dl, size_t buf_len)
+{
+	char *p;
+	char *q;
+	size_t parse_len;
+	unsigned int expected_status;
+	struct transport_params_http *http;
+
+	http = (struct transport_params_http *)dl->transport_internal;
+
+	LOG_DBG("(partial) http header response:\n%s", dl->config.buf);
+
+	p = strnstr(dl->config.buf, "\r\n\r\n", dl->config.buf_size);
+	if (p) {
+		/* End of header received */
+		http->header.has_end = true;
+		parse_len = p + strlen("\r\n\r\n") - (char *)dl->config.buf;
+	} else {
+		parse_len = buf_len;
+	}
+
+	for (size_t i = 0; i < parse_len; i++) {
+		dl->config.buf[i] = tolower(dl->config.buf[i]);
+	}
+
+	/* Look for the status code just after "http/1.1 " */
+	p = strnstr(dl->config.buf, "http/1.1 ", parse_len);
+	if (p) {
+		q = strnstr(p, "\r\n", parse_len - (p - dl->config.buf));
+		if (q) {
+			/* Received entire line */
+			p += strlen("http/1.1 ");
+			http->header.status_code = strtoul(p, &q, 10);
+		}
+
+	}
+
+	/* The file size is returned via "Content-Length" in case of HTTP,
+	 * and via "Content-Range" in case of HTTPS with range requests.
+	 */
+	do {
+		if (dl->file_size == 0) {
+			if (http->ranged) {
+				p = strnstr(dl->config.buf, "\r\ncontent-range", parse_len);
+				if (!p) {
+					break;
+				}
+				p = strnstr(p, "/", parse_len - (p - dl->config.buf));
+				if (!p) {
+					break;
+				}
+				q = strnstr(p, "\r\n", parse_len - (p - dl->config.buf));
+				if (!q) {
+					/* Missing end of line */
+					break;
+				}
+			} else { /* proto == PROTO_HTTP */
+				p = strnstr(dl->config.buf, "\r\ncontent-length", parse_len);
+				if (!p) {
+					break;
+				}
+				p = strstr(p, ":");
+				if (!p) {
+					break;
+				}
+				q = strnstr(p, "\r\n", parse_len - (p - dl->config.buf));
+				if (!q) {
+					/* Missing end of line */
+					break;
+				}
+				/* Accumulate any eventual progress (starting offset)
+				 * when reading the file size from Content-Length
+				 */
+				dl->file_size = dl->progress;
+			}
+
+			dl->file_size += atoi(p + 1);
+			LOG_DBG("File size = %u", dl->file_size);
+		}
+	} while (0);
+
+	p = strnstr(dl->config.buf, "\r\nconnection: close", parse_len);
+	if (p) {
+		LOG_WRN("Peer closed connection, will re-connect");
+		http->connection_close = true;
+	}
+
+	if (http->header.has_end) {
+		/* We have received the end of the header.
+		 * Verify that we have received everything that we need.
+		 */
+
+		if (!http->header.status_code) {
+			LOG_ERR("Server response malformed: status code not found");
+			return -EBADMSG;
+		}
+
+		expected_status = (http->ranged || dl->progress) ? 206 : 200;
```

Comment:
These two http status codes can be defined too, mostly for their name/use. For example, I know from the top of my head that 200 is success, but don't remember 206.